### PR TITLE
Render Passes

### PR DESCRIPTION
**Pull request by kainino0x**
There's a lot here. I think it's worth a first pass review finally. PTAL

EDIT: For TODOs, see Issue #7.

### CODE REVIEW

**Comment by Kangz** [comment]
FYI I merged the errors PR so you'll need to do a couple changes:
 - Builders inherit from Builder<ObjectBase>
 - Builders GetResult becomes GetResultImpl() overrride and is private
 - Most cases of device->HandleError become HandleError

File: next.json [hunk 1]
```diff
@@ -493,6 +547,42 @@
...
+            {
+                "name": "enable attachment",
+                "comment": "Lazy implementation; probably should make an Attachment object.",
+                "args": [
+                    {"name": "attachment slot", "type": "uint32_t"},
+                    {"name": "format", "type": "texture format"}
+                ]
+            },
+            {
+                "name": "set subpass count",
```
Comment (Kangz) [inline next.json]: Likewise we should probably have a "set attachment count" and validate that both are done before anything else?

File: src/backend/common/CommandBuffer.cpp [hunk 2]
```diff
@@ -238,9 +257,75 @@ namespace backend {
...
+                case Command::BeginRenderPass:
+                    {
+                        BeginRenderPassCmd* begin = iterator.NextCommand<BeginRenderPassCmd>();
+                        auto* renderPass = begin->renderPass.Get();
+                        auto* framebuffer = begin->framebuffer.Get();
+                        if (currentRenderPass != nullptr) {
+                            device->HandleError("A render pass is already active");
+                            return false;
+                        }
+                        if (framebuffer->GetRenderPass() != renderPass) {
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: Leave a TODO that this compatibility check will need to be slightly relaxed at some point, to allow render passes with different load and store ops.

File: src/backend/common/CommandBuffer.cpp [hunk 3]
```diff
@@ -238,9 +257,75 @@ namespace backend {
...
+                            device->HandleError("Framebuffer is incompatible with this render pass");
+                            return false;
+                        }
+                        lastPipeline = nullptr;
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: What the reason to reset the pipeline and bindgroups? I think there could be a use for inheriting bindgroups from outside the pipeline.

I'm not sure what the best way to go is, but this is inconsistent with endrenderpass that also reset vertex and index buffers.

Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: Also we could add a validation aspect for renderpasses.

File: src/backend/common/CommandBuffer.cpp [hunk 4]
```diff
@@ -358,21 +443,44 @@ namespace backend {
                     }
                     break;
 
+                case Command::EndRenderPass:
+                    {
+                        iterator.NextCommand<EndRenderPassCmd>();
+                        if (currentRenderPass == nullptr) {
+                            device->HandleError("No render pass is currently active");
+                            return false;
+                        }
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: We should also check that we are at the last subpass.

File: src/backend/common/CommandBuffer.cpp [hunk 5]
```diff
@@ -358,21 +443,44 @@ namespace backend {
...
-                            aspects.reset(VALIDATION_ASPECT_RENDER_PIPELINE);
                         } else {
+                            if (currentRenderPass == nullptr) {
+                                device->HandleError("A render pass must be active when a render pipeline is set");
+                                return false;
+                            }
                             aspects.set(VALIDATION_ASPECT_RENDER_PIPELINE);
-                            aspects.reset(VALIDATION_ASPECT_COMPUTE_PIPELINE);
                         }
-                        aspects.reset(VALIDATION_ASPECT_BINDGROUPS);
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: This dirties the validation of vertex buffers and bindgroups so that we can perform it again at draw time. It should be kept.

File: src/backend/common/CommandBuffer.cpp [hunk 6]
```diff
@@ -476,6 +584,15 @@ namespace backend {
                             return false;
                         }
 
+                        auto& subpassInfo = currentRenderPass->GetSubpassInfo(currentSubpass);
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: In the object introduced to handle all the usage tracking, this could be done with a "temporarily frozen" tag put on resources.

File: src/backend/common/Forward.h [hunk 7]
```diff
@@ -64,6 +68,9 @@ namespace backend {
     static constexpr uint32_t kMaxVertexAttributes = 16u;
     static constexpr uint32_t kMaxVertexInputs = 16u;
     static constexpr uint32_t kNumStages = 3;
+    static constexpr uint32_t kMaxTextureViewsPerFramebuffer = 16u;
```
Comment (Kangz) [inline src/backend/common/Forward.h]: Is there a practical reason for this, like a Vulkan limitation? Otherwise Framebuffer could require the renderpass to be given before the texture views, and resize a vector.

File: src/backend/common/Forward.h [hunk 8]
```diff
@@ -64,6 +68,9 @@ namespace backend {
...
+    static constexpr uint32_t kMaxRenderPassAttachments = 16u;
```
Comment (Kangz) [inline src/backend/common/Forward.h]: ditto, but for renderpasses

File: src/backend/common/Framebuffer.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    bool FramebufferBuilder::WasConsumed() const {
+        return consumed;
+    }
+
+    FramebufferBase* FramebufferBuilder::GetResult() {
+        constexpr int allProperties = FRAMEBUFFER_PROPERTY_RENDER_PASS | FRAMEBUFFER_PROPERTY_DIMENSIONS;
+        if ((propertiesSet & allProperties) != allProperties) {
+            device->HandleError("Framebuffer missing properties");
+            return nullptr;
+        }
```
Comment (Kangz) [inline src/backend/common/Framebuffer.cpp]: Also needs to check all attachments required by the renderpass are set, that is, that no element of the vector is null.

File: src/backend/common/Framebuffer.h [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        private:
+            friend class FramebufferBase;
+
+            DeviceBase* device = nullptr;
+            Ref<RenderPassBase> renderPass;
+            uint32_t width = 0;
+            uint32_t height = 0;
+            std::array<Ref<TextureViewBase>, kMaxTextureViewsPerFramebuffer> textureViews;
+            int propertiesSet = 0;
+            bool consumed = false;
```
Comment (Kangz) [inline src/backend/common/Framebuffer.h]: Also with the merge of the errors PR, all the "consumed" disappeared

File: src/backend/common/RenderPass.h [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+#include <vector>
+
+namespace backend {
+
+    class RenderPassBase : public RefCounted {
+        public:
+            RenderPassBase(RenderPassBuilder* builder);
+
+            struct SubpassInfo {
+                std::map<uint32_t, uint32_t> colorAttachments;
```
Comment (Kangz) [inline src/backend/common/RenderPass.h]: std::array<kMaxColorAttachments, uint32_t>
will need to update the code in CommandBuffer that iterates over it to check for nulltpr

File: src/backend/common/RenderPass.cpp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    void RenderPassBuilder::SetSubpassColorAttachment(uint32_t subpassIndex, uint32_t outputAttachmentLocation, uint32_t attachmentSlot) {
+        if (subpassIndex >= subpasses.size()) {
+            device->HandleError("Subpass index out of bounds");
+            return;
+        }
+        if (outputAttachmentLocation >= kMaxColorAttachments) {
+            device->HandleError("Output attachment location out of bounds");
+            return;
+        }
+        if (!attachmentsEnabled[attachmentSlot]) {
```
Comment (Kangz) [inline src/backend/common/RenderPass.cpp]: I think we can force attachments to be a full and contiguous range so this check disappears.

File: src/backend/common/RenderPass.cpp [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            device->HandleError("Attachment slot not enabled");
+            return;
+        }
+
```
Comment (Kangz) [inline src/backend/common/RenderPass.cpp]: Also check we are not setting it a second time.

File: src/backend/common/RenderPass.cpp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    RenderPassBuilder::RenderPassBuilder(DeviceBase* device)
+        : device(device), subpasses(1) {
+    }
+
+    bool RenderPassBuilder::WasConsumed() const {
+        return consumed;
+    }
+
+    RenderPassBase* RenderPassBuilder::GetResult() {
+        consumed = true;
```
Comment (Kangz) [inline src/backend/common/RenderPass.cpp]: We'll need to check a whole lot of things here once we add preserved attachments, and stuff like that.

But also for now we should check that all the attachments and subpasses have been defined (I don't see a usecase for sparse index spaces).

File: src/backend/common/CommandBuffer.cpp [hunk 15]
```diff
@@ -238,9 +257,75 @@ namespace backend {
...
+                        currentFramebuffer = framebuffer;
+                        currentSubpass = 0;
+                        if (!beginSubpass()) {
+                            return false;
+                        }
+                    }
+                    break;
+
                 case Command::CopyBufferToTexture:
                     {
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: Blits should be forbidden inside renderpasses

File: src/backend/common/Framebuffer.cpp (hunk 9 above)
Comment (Kangz) [inline src/backend/common/Framebuffer.cpp]: Or count that there is at most one unset attachment for the "default framebuffer" hack and add a TODO ;)

File: src/backend/common/RenderPass.h (hunk 11 above)
Comment (kainino0x) [inline src/backend/common/RenderPass.h]: Had to also add a `std::bitset<kMaxColorAttachments> colorAttachmentsSet` - I think this is reasonable.

File: src/backend/common/RenderPass.cpp (hunk 14 above)
Comment (kainino0x) [inline src/backend/common/RenderPass.cpp]: This is sort of a fuzzy concept as these could have optional properties with default values, but I've added this validation for the format field of attachments at least. A subpass by default will just not have any attachments.

File: next.json [hunk 16]
```diff
@@ -493,6 +547,42 @@
...
+                "comment": "Lazy implementation; probably should make a Subpass object.",
+                "args": [
+                    {"name": "subpass count", "type": "uint32_t"}
+                ]
+            },
+            {
+                "name": "set subpass color attachment",
+                "comment": "Lazy implementation; probably should make a Subpass object.",
```
Comment (Kangz) [inline next.json]: You mean as in not having a dedicated subpass object? Sounds good.

File: src/backend/common/CommandBuffer.cpp [hunk 17]
```diff
@@ -238,9 +257,75 @@ namespace backend {
...
         };
 
+        // TODO(kainino@chromium.org): Manage this state inside an object, change lambda to a method
+        RenderPassBase* currentRenderPass = nullptr;
+        FramebufferBase* currentFramebuffer = nullptr;
+        uint32_t currentSubpass = 0;
+        auto beginSubpass = [&]() -> bool {
+            auto& subpassInfo = currentRenderPass->GetSubpassInfo(currentSubpass);
+            for (auto& kv : subpassInfo.colorAttachments) {
+                if (auto* tv = currentFramebuffer->GetTextureView(kv.second)) {
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: Yep realized that afterwards, but forgot to edit this comment, TODO sounds good.

File: src/backend/common/RenderPass.h (hunk 11 above)
Comment (Kangz) [inline src/backend/common/RenderPass.h]: Bitset sounds good.

File: examples/HelloVertices.cpp [hunk 18]
```diff
@@ -80,8 +80,9 @@ void init() {
 void frame() {
     static const uint32_t vertexBufferOffsets[1] = {0};
     nxt::RenderPass renderpass = device.CreateRenderPassBuilder()
-        .EnableAttachment(0, nxt::TextureFormat::R8G8B8A8Unorm)
-        .SetSubpassColorAttachment(0, 0, 0)
+        .SetAttachmentCount(1)
+        .AttachmentSetFormat(0, nxt::TextureFormat::R8G8B8A8Unorm)
+        .SubpassSetColorAttachment(0, 0, 0)
```
Comment (Kangz) [inline examples/HelloVertices.cpp]: There is a missing SetSubpassCount here, we should double check validation catches this case.

File: next.json [hunk 19]
```diff
@@ -423,11 +423,16 @@
                 ]
             },
             {
-                "name": "set attachments",
+                "name": "set attachment count",
```
Comment (Kangz) [inline next.json]: The attachment count can implicitly be the same one as the renderpass.

File: src/backend/common/Framebuffer.cpp [hunk 20]
```diff
@@ -52,12 +53,28 @@ namespace backend {
...
+                if (textureView == nullptr) {
+                    if (usingBackbufferHack) {
+                        device->HandleError("Some subpass uses more than one texture view which is null in this framebuffer");
+                        return nullptr;
+                    }
+                    usingBackbufferHack = true;
+                }
+            }
+        }
+
```
Comment (Kangz) [inline src/backend/common/Framebuffer.cpp]: nit: check format and samples match the renderpass.

File: src/backend/common/CommandBuffer.cpp [hunk 21]
```diff
@@ -597,14 +598,25 @@ namespace backend {
...
 
-                        // TODO(kainino@chromium.org): We should be able to optimize this by tracking which textures are in use as attachments.
-                        auto& subpassInfo = currentRenderPass->GetSubpassInfo(currentSubpass);
-                        for (auto attachmentSlot : subpassInfo.colorAttachments) {
-                            // XXX: this loop is running too long; how do we make it work correctly?
-                            auto* tv = currentFramebuffer->GetTextureView(attachmentSlot);
-                            if (tv && tv->GetTexture() == texture) {
-                                device->HandleError("Can't transition a texture while it's used as a color attachment");
-                                return false;
+                        // TODO(kainino@chromium.org): We should be able to
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: I'm thinking that we could have scratch buffers in textures and buffers, that can be used during validation to store custom data such as this. Validation looks good like this for now.

File: src/backend/common/RenderPass.cpp (hunk 14 above)
Comment (Kangz) [inline src/backend/common/RenderPass.cpp]: Yeah this is a bit fuzzy, I'll think about this more.

File: src/backend/common/RenderPass.cpp (hunk 13 above)
Comment (Kangz) [inline src/backend/common/RenderPass.cpp]: I don't see the bitset being checked.

**Comment by Kangz** [comment]
Added a couple comments but reviewing the incremental changes is a bit hard, I'll do a full re-review too.

File: examples/HelloVertices.cpp (hunk 18 above)
Comment (kainino0x) [inline examples/HelloVertices.cpp]: Currently, subpass count defaults to 1, attachment count defaults to 0. They aren't required.

File: next.json (hunk 19 above)
Comment (kainino0x) [inline next.json]: Thanks, good point.

File: src/backend/common/Framebuffer.cpp (hunk 20 above)
Comment (kainino0x) [inline src/backend/common/Framebuffer.cpp]: Added this validation in SetAttachment

File: src/backend/common/RenderPass.cpp (hunk 12 above)
Comment (kainino0x) [inline src/backend/common/RenderPass.cpp]: Done

File: src/backend/common/RenderPass.cpp (hunk 13 above)
Comment (kainino0x) [inline src/backend/common/RenderPass.cpp]: It's there for me, this diff is outdated (ergh github...)

**Comment by kainino0x** [comment]
Sorry, I take that back, there's still one more thing I need to do (that I forgot about after the weekend).

**Comment by kainino0x** [comment]
Okay, fixed some bugs and update the examples. Now it's ready.

BTW I'm having some issues with error handling - most API usage errors (HandleError) seem to cause segfaults due to dereferencing null built objects. Maybe there's an issue in my code, maybe in the error handling code, not sure.

File: src/backend/common/Pipeline.cpp [hunk 22]
```diff
@@ -33,6 +35,12 @@ namespace backend {
             return;
         }
 
+        if (!IsCompute() && !renderPass) {
+            builder->HandleError("Pipeline render pass not set");
+            return;
+        }
+        // TODO(kainino@chromium.org): Need to verify the pipeline against its render subpass.
```
Comment (kainino0x) [inline src/backend/common/Pipeline.cpp]: This is a bit of work because the info is (afaik) only available in ExtractSpirvInfo, so I'm going to make it a separate patch.

File: src/backend/common/CommandBuffer.cpp [hunk 23]
```diff
@@ -472,6 +594,28 @@ namespace backend {
                             return false;
                         }
 
+                        // TODO(kainino@chromium.org): We should be able to
+                        // optimize this by tracking which textures are in use
+                        // as attachments. Maybe it's possible that the
+                        // XAttachment usages could mark this: disallow
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: note: This could work and this would mean the end usage will have to be encoded in the renderpass. I'm not 100% sure this is desirable but don't see a counter example.

File: src/backend/common/CommandBuffer.cpp [hunk 24]
```diff
@@ -472,6 +594,28 @@ namespace backend {
...
+                        // in an XAttachment usage.
+                        if (currentRenderPass) {
+                            auto& info = currentRenderPass->GetSubpassInfo(currentSubpass);
+                            for (uint32_t location = 0; location < info.colorAttachments.size(); ++location) {
+                                if (!info.colorAttachmentsSet[location]) {
+                                    continue;
+                                }
+                                uint32_t attachmentSlot = info.colorAttachments[location];
+                                auto* tv = currentFramebuffer->GetTextureView(attachmentSlot);
+                                if (tv && tv->GetTexture() == texture) {
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: nit: TODO, check for tv != nullptr will not be needed once we have WSI

File: src/backend/common/Framebuffer.cpp [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+            return;
+        }
+
+        this->width = width;
+        this->height = height;
+        propertiesSet |= FRAMEBUFFER_PROPERTY_DIMENSIONS;
+    }
+
+    void FramebufferBuilder::SetAttachment(uint32_t attachmentSlot, TextureViewBase* textureView) {
+        if ((propertiesSet & FRAMEBUFFER_PROPERTY_RENDER_PASS) != FRAMEBUFFER_PROPERTY_RENDER_PASS) {
```
Comment (Kangz) [inline src/backend/common/Framebuffer.cpp]: nit: properties & RENDER_PASS == 0

File: src/backend/common/Framebuffer.cpp [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    FramebufferBase* FramebufferBuilder::GetResultImpl() {
+        constexpr int requiredProperties = FRAMEBUFFER_PROPERTY_RENDER_PASS | FRAMEBUFFER_PROPERTY_DIMENSIONS;
+        if ((propertiesSet & requiredProperties) != requiredProperties) {
+            HandleError("Framebuffer missing properties");
+            return nullptr;
+        }
+
+        for (uint32_t subpass = 0; subpass < renderPass->GetSubpassCount(); ++subpass) {
```
Comment (Kangz) [inline src/backend/common/Framebuffer.cpp]: Since the renderpass's attachments must all be defined, we only need to iterate over this->textureViews.

File: src/backend/common/Pipeline.cpp (hunk 22 above)
Comment (Kangz) [inline src/backend/common/Pipeline.cpp]: Souds good

File: src/backend/common/CommandBuffer.cpp [hunk 27]
```diff
@@ -354,18 +452,42 @@ namespace backend {
...
+                                return false;
+                            }
                             aspects.set(VALIDATION_ASPECT_COMPUTE_PIPELINE);
-                            aspects.reset(VALIDATION_ASPECT_RENDER_PIPELINE);
                         } else {
+                            if (!currentRenderPass) {
+                                device->HandleError("A render pass must be active when a render pipeline is set");
+                                return false;
+                            }
                             aspects.set(VALIDATION_ASPECT_RENDER_PIPELINE);
```
Comment (Kangz) [inline src/backend/common/CommandBuffer.cpp]: nit: need to check that the pipeline's renderpass is compatible with currentRenderpass.

File: src/backend/common/RenderPass.cpp [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+    enum RenderPassSetProperties {
+        RENDERPASS_PROPERTY_ATTACHMENT_COUNT = 0x1,
+        RENDERPASS_PROPERTY_SUBPASS_COUNT = 0x2,
+    };
+
+    RenderPassBuilder::RenderPassBuilder(DeviceBase* device)
+        : Builder(device), subpasses(1) {
+    }
+
+    RenderPassBase* RenderPassBuilder::GetResultImpl() {
```
Comment (Kangz) [inline src/backend/common/RenderPass.cpp]: check that both renderpass count and attachment count are set. This will trigger errors in some of the samples.

Comment (Kangz) [inline src/backend/common/RenderPass.cpp]: check that both the subpass count and the attachment counts are set. Some samples will need to be fixed.

File: src/backend/common/RenderPass.h [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+        private:
+            friend class RenderPassBase;
+
+            enum AttachmentProperty {
+                ATTACHMENT_PROPERTY_FORMAT,
+                ATTACHMENT_PROPERTY_COUNT
+            };
+
+            enum SubpassProperty {
+                SUBPASS_PROPERTY_COUNT
```
Comment (Kangz) [inline src/backend/common/RenderPass.h]: Do you expect many properties for subpasses and attachments? I'm wondering if just a std::vector<bool> would be enough.

File: examples/Animometer.cpp [hunk 30]
```diff
@@ -22,6 +22,8 @@
 nxt::Device device;
 nxt::Queue queue;
 nxt::Pipeline pipeline;
+nxt::RenderPass renderpass;
```
Comment (Kangz) [inline examples/Animometer.cpp]: nit: maybe these "default" renderpass and FBO could be put in Utils

File: examples/Animometer.cpp [hunk 31]
```diff
@@ -103,7 +105,17 @@ void init() {
         })"
     );
 
+    renderpass = device.CreateRenderPassBuilder()
+        .SetAttachmentCount(1)
+        .AttachmentSetFormat(0, nxt::TextureFormat::R8G8B8A8Unorm)
```
Comment (Kangz) [inline examples/Animometer.cpp]: nit: needs subpass count

File: examples/ComputeBoids.cpp [hunk 32]
```diff
@@ -134,7 +136,17 @@ void initRender() {
         .SetInput(1, sizeof(glm::vec2), nxt::InputStepMode::Vertex)
         .GetResult();
 
+    renderpass = device.CreateRenderPassBuilder()
+        .SetAttachmentCount(1)
+        .AttachmentSetFormat(0, nxt::TextureFormat::R8G8B8A8Unorm)
+        .SubpassSetColorAttachment(0, 0, 0)
```
Comment (Kangz) [inline examples/ComputeBoids.cpp]: ditto

File: examples/HelloCompute.cpp [hunk 33]
```diff
@@ -107,7 +109,17 @@ void init() {
             .SetBindGroupLayout(0, bgl)
             .GetResult();
 
+        renderpass = device.CreateRenderPassBuilder()
+            .SetAttachmentCount(1)
+            .AttachmentSetFormat(0, nxt::TextureFormat::R8G8B8A8Unorm)
```
Comment (Kangz) [inline examples/HelloCompute.cpp]: ditto, and etc in other samples

File: src/backend/common/RenderPass.cpp (hunk 28 above)
Comment (kainino0x) [inline src/backend/common/RenderPass.cpp]: TBD

File: src/backend/common/RenderPass.h (hunk 29 above)
Comment (Kangz) [inline src/backend/common/RenderPass.h]: I meant ```std::vector<bool>``` for some reason GitHub ate the template param. It sounds that most subpass things could have obvious defaults, but not attachments, so AttachmentProperty looks needed.

Thanks!

Comment (kainino0x) [inline src/backend/common/RenderPass.h]: Oh, I see re:`<bool>`.

I don't think defaults for subpass attachments make that much sense actually, but if we want to track them it's going to be more complex than a vector of bitsets, so I've removed SubpassProperty.

File: examples/ComputeBoids.cpp (hunk 32 above)
Comment (kainino0x) [inline examples/ComputeBoids.cpp]: Done

File: examples/HelloCompute.cpp (hunk 33 above)
Comment (kainino0x) [inline examples/HelloCompute.cpp]: Done

File: src/backend/common/RenderPass.cpp (hunk 28 above)
Comment (kainino0x) [inline src/backend/common/RenderPass.cpp]: Done

File: examples/Animometer.cpp (hunk 30 above)
Comment (kainino0x) [inline examples/Animometer.cpp]: Good idea, I'll do this

Comment (kainino0x) [inline examples/Animometer.cpp]: I put just the setup in Utils `CreateDefaultRenderPass` to initialize a render pass and framebuffer.

**Comment by kainino0x** [comment]
Addressed comments and also squashed everything since I have no idea how well GitHub handles squashing merge commits.

File: src/backend/common/CommandBuffer.cpp (hunk 23 above)
Comment (kainino0x) [inline src/backend/common/CommandBuffer.cpp]: I was thinking of disallowing explicit transitions INTO XAttachment, but having render passes always leave their attachments in the 'no usage' state (if they're not frozen). Then we don't have to set the end usage in the render pass.

Comment (kainino0x) [inline src/backend/common/CommandBuffer.cpp]: (Or leaving their attachments in the 'XAttachment' usage, because it doesn't matter whether it's that or 'no usage', I think.)

**Comment by Kangz** [comment]
Just to be sure, do you want to merge this now or would you rather wait until the OpenGL backend is done? (and potentially other checkboxes)

**Comment by kainino0x** [comment]
Hm, I think it would be better to merge now. I don't want to let it diverge too much, and it's going to take me a while to do the remaining tasks (been focusing on Chrome for the past few days).

**Comment by Kangz** [comment]
LGTM, an ultra-nit is that "Squashed commit of ..." isn't the nicest. Maybe it could be "introduce renderpasses" instead?

**Comment by kainino0x** [comment]
Rewrote commit message

**Comment by Kangz** [comment]
LGTM
